File: LeetCodes(Daily_Practice)/19_Remove_Nth_Node_From_End_of_List.go
package main

import "fmt"

type node struct {
	data int
	next *node
}

type linkedList struct {
	head   *node
	tail   *node
	length int
}

func (l *linkedList) len() {
	fmt.Println("Size of list ", l.length)
}

func (l *linkedList) addlast(value int) {
	newNode := node{data: value}
	p := l.head
	count := 0
	if l.head == nil {
		l.head = &newNode

	} else {
		for count < l.length-1 {
			p = p.next
			count++
		}
		p.next = &newNode
	}
	l.tail = &newNode
	l.length++
}

func (l *linkedList) printList() {
	if l.head == nil {
		return
	}
	currentNode := l.head
	for currentNode != nil {
		fmt.Print(currentNode.data, ", ")
		currentNode = currentNode.next
	}
	fmt.Println()
}

// Removing given positions element from lastside of Linkedlist.
func (l *linkedList) removeElement(position int) {
	if l.head == nil {
		return
	}
	x := l.length - position
	p := l.head
	count := 0

	for count < x-1 {
		p = p.next
		count++
	}
	ele := p.next.data
	p.next = p.next.next
	l.length--
	fmt.Println(ele, "deleted")
}

func main() {
	myList := linkedList{}
	myList.addlast(89)
	myList.addlast(100)
	myList.addlast(201)
	myList.addlast(488)
	myList.addlast(167)
	myList.addlast(983)
	myList.printList()
	myList.len()
	myList.removeElement(4)
	myList.printList()
	myList.len()
}

File: Daily_Practice/19_Remove_Nth_Node_From_End_of_List.py
class _Node:
    __slots__ = '_element', '_next'

    def __init__(self, element, next):
        self._element = element
        self._next = next

class LinkedList:
    def __init__(self):
        self._head = None
        self._tail = None
        self._size = 0

    def __len__(self):
        return self._size
    
    def isempty(self):
        return self._size == 0
    
    def addlast(self, e):
        new = _Node(e, None)
        if self.isempty():
            self._head = new
        else:
            self._tail._next = new
        self._tail = new
        self._size += 1
    # To remove nth node from last side of list.
    def remove(self, e):
        if self.isempty():
            print("List is empty..")
            return
        p = self._head
        x = self._size - e
        count = 0
        while count < x-1:
            p = p._next
            count += 1
        d = p._next._element
        p._next = p._next._next
        print(d,"deleted")

    def display(self):
        p = self._head
        while p:
            print(p._element, end=", ")
            p = p._next
        print()

    def rotateRight(self, head, k):
        if not head:
            return head

        #connect tail to head
        cur= self._head
        length =1
        while cur._next:
            cur = cur._next
            length+=1 
        cur._next = self._head

        #move to new head
        k= length - (k%length)
        while k>0:
            cur=cur._next
            k-=1

        #disconnect and return new head
        newhead = cur._next
        cur._next = None
        return newhead

obj = LinkedList()
obj.addlast(9)
obj.addlast(2)
obj.addlast(3)
obj.addlast(5)
obj.addlast(7)
obj.display()
obj.remove(1)
obj.display()
tp = obj
op = obj.rotateRight(tp, 2)
print(op)